### DISCUSSION TITLE
Styling individual chips

### DISCUSSION THREAD

**Discussion by janpeterka** [josefarias/hotwire_combobox#168] [Ideas]
I'm creating filter with multiple multiselect inputs (recipe category, dietary labels,...), and want to show all selected in one div - see image:
![Screenshot from 2024-05-19 22-49-28](https://github.com/josefarias/hotwire_combobox/assets/6343648/a7d6cde7-c1bb-406a-ab17-d43076612c4e)

First part - showing them in same div - can be done setting `param[:for_id]`, no problem here (it's not documented, and it cannot be set in `multiselect_chip_src`, but other than that, it's ok).

But I also want to separate them by color. And it seems (looking into code for `hw_combobox_selection_chips_for`) not possible to style chips.

Am I missing something?

**Comment by janpeterka** [comment]
Ok, after some more trying to make this work I found out that this case is not the best as it doesn't really work with `remove` action (as tags are outside the input div). I'll try some prodding into that too, but option to style chips is still relevant to me.

**Comment by josefarias** [comment]
Hey @janpeterka thanks for the question.

As you've noticed, rendering the chips in a separate container will likely cause issues downstream. I don't have plans to support that directly unfortunately.

However, the good news is you can absolutely style chips yourself! It's a matter of implementing your own turbo stream templates for chips. Check out this example from the system tests:

1. Set the source

https://github.com/josefarias/hotwire_combobox/blob/ce2bfbb3cfbbc0d602eef8c126d7ec0d8b417e0b/test/dummy/app/views/comboboxes/multiselect_async_html.html.erb#L46-L51

2. Implement controller action

https://github.com/josefarias/hotwire_combobox/blob/ce2bfbb3cfbbc0d602eef8c126d7ec0d8b417e0b/test/dummy/app/controllers/state_chips_controller.rb#L8-L9

3. Create your own chip tempalte

https://github.com/josefarias/hotwire_combobox/blob/ce2bfbb3cfbbc0d602eef8c126d7ec0d8b417e0b/test/dummy/app/views/state_chips/create_html.turbo_stream.erb#L1-L9

You're right that this isn't documented yet. I'm waiting to see what kind of issues come up before settling on an API.

Let me know if that doesn't work for you. Closing in the meantime.

**Comment by janpeterka** [comment]
Hi @josefarias, thank you for answering!

> As you've noticed, rendering the chips in a separate container will likely cause issues downstream. I don't have plans to support that directly unfortunately.

I was thinking about looking into that possibly (maybe using some separate controller with outlets or something), would you accept it then, or you don't want to complicate things?
I will probably find some simpler UI for now, as what I imagined doesn't really have simple solution.

> However, the good news is you can absolutely style chips yourself! It's a matter of implementing your own turbo stream templates for chips.

Ok, it's lot simpler than I though, thanks!. But still, maybe default helper/template could support `class:` parameter?

**Reply by josefarias** [reply]
> I was thinking about looking into that possibly (maybe using some separate controller with outlets or something), would you accept it then, or you don't want to complicate things?

Probably not, I'm afraid. I'm open to being talked into it but I'm mostly interested in providing _a_ solution (as opposed to _every_ solution)

> maybe default helper/template could support class: parameter?

I'd say not right now. If adding a class is all you need right now, then it should be pretty simple to copy the original chip template and go from there: https://github.com/josefarias/hotwire_combobox/blob/d9c1e1fc3522faa6f149e5f07527cfaa645c541f/app/views/hotwire_combobox/_selection_chip.turbo_stream.erb#L3-L8

**Reply by janpeterka** [reply]
Ok, understood :))

I'll try to write my custom chip, possibly with custom controller and if I succeed, I'll propose to add it to examples, or at least to discussions :)

Thanks for your work, this gem really is great!

**Reply by josefarias** [reply]
Sounds great! Appreciate your understanding. I'm wary of saying yes to too many things this early.

Thanks for raising these points!
